### PR TITLE
[Docs] Update Train user guide to use the new APIs

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Revamps the Train user guide to use the new APIs.

WIP. Depends on https://github.com/ray-project/ray/pull/25735

Will move to doc code in a followup PR to simplify review.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Yard1** [comment]
cc @amogkam @matthewdeng

File: doc/source/train/user_guide.rst [hunk 1]
```diff
@@ -224,31 +231,40 @@ To customize the ``backend`` setup, you can replace the string argument with a
...
 
-        from ray.train import Trainer
-        from ray.train.torch import TorchConfig
+        from ray.train.torch import TorchTrainer, TorchConfig
 
-        trainer = Trainer(backend=TorchConfig(...), num_workers=2)
+        trainer = TorchTrainer(
+            train_func,
+            torch_backend=TorchConfig(...),
+            scaling_config=dict(num_workers=2),
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:239]: cc @sumanthratna we'll need to be aware of updating this to use the `ScalingConfig` dataclass

File: doc/source/train/user_guide.rst [hunk 2]
```diff
@@ -182,40 +183,46 @@ training.
...
         # For GPU Training, set `use_gpu` to True.
-        # trainer = Trainer(backend="horovod", num_workers=2, use_gpu=True)
+        use_gpu = False
+        trainer = HorovodTrainer(
+            train_func,
+            scaling_config=dict(use_gpu=use_gpu, num_workers=2)
+        )
 
 To customize the ``backend`` setup, you can replace the string argument with a
 :ref:`train-api-backend-config` object.
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:228]: Update this since there is no longer a string argument

File: doc/source/train/user_guide.rst [hunk 3]
```diff
@@ -258,85 +274,77 @@ ready to start training!
...
 A primary use-case for ``config`` is to try different hyperparameters. To
 perform hyperparameter tuning with Ray Train, please refer to the
 :ref:`Ray Tune integration <train-tune>`.
 
 .. TODO add support for with_parameters
 
+.. _train-result-object:
 
-.. _train-log-dir:
+The Result Object
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:332]: Can we add a TODO here to flesh this section out? Feels like we should have some AIR documentation that describes how to read results

File: doc/source/ray-air/package-ref.rst [hunk 4]
```diff
@@ -124,4 +124,16 @@ Configs
...
+.. _air-builtin-callbacks:
+
+Callbacks
+~~~~~~~~~
+
+Keras
+#####
+
+.. automodule:: ray.air.callbacks.keras
+    :members:
```
Comment (matthewdeng) [inline doc/source/ray-air/package-ref.rst:139]: This isn't loading for me

File: doc/source/train/user_guide.rst [hunk 5]
```diff
@@ -369,45 +375,7 @@ the end of each training epoch.
...
-    _date
-    # The worker hostname (platform.node()).
-    _hostname
-    # The worker IP address.
-    _node_ip
-    # The worker process ID (os.getpid()).
-    _pid
-    # The cumulative training time of all iterations so far.
-    _time_total_s
+            session.report({"results": results})
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:378]: Add import for `session`

File: doc/source/train/user_guide.rst [hunk 6]
```diff
@@ -346,20 +354,18 @@ Logging, Monitoring, and Callbacks
 ----------------------------------
 
 Ray Train has mechanisms to easily collect intermediate results from the training workers during the training run
-and also has a :ref:`Callback interface <train-callbacks>` to perform actions on these intermediate results (such as logging, aggregations, printing, etc.).
-You can use either the :ref:`built-in callbacks <train-builtin-callbacks>` that Ray Train provides,
-or implement a :ref:`custom callback <train-custom-callbacks>` for your use case.
+and also has a :ref:`Callback interface <train-callbacks>` to perform actions on these intermediate results (such as logging, aggregations, etc.).
+You can use either the :ref:`built-in callbacks <air-builtin-callbacks>` that Ray AIR provides,
+or implement a :ref:`custom callback <train-custom-callbacks>` for your use case. The callback API
+is shared with Ray Tune.
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:360]: Outside of the scope of this PR, but we should figure out how to cleanly separate the docs for Train/Tune/AIR. I think for a new user trying out Train the dependencies on Tune is a bit confusing.

File: doc/source/train/user_guide.rst [hunk 7]
```diff
@@ -449,105 +404,82 @@ Example: Logging to MLflow and TensorBoard
...
-    # [{'epoch': 0, '_timestamp': 1630471763, '_time_this_iter_s': 0.0020279884338378906, '_training_iteration': 1}, {'epoch': 0, '_timestamp': 1630471763, '_time_this_iter_s': 0.0014922618865966797, '_training_iteration': 1}]
-    # [{'epoch': 1, '_timestamp': 1630471763, '_time_this_iter_s': 0.0008401870727539062, '_training_iteration': 2}, {'epoch': 1, '_timestamp': 1630471763, '_time_this_iter_s': 0.0007486343383789062, '_training_iteration': 2}]
-    # [{'epoch': 2, '_timestamp': 1630471763, '_time_this_iter_s': 0.0014500617980957031, '_training_iteration': 3}, {'epoch': 2, '_timestamp': 1630471763, '_time_this_iter_s': 0.0015292167663574219, '_training_iteration': 3}]
-    trainer.shutdown()
+    trainer.fit()
 
-..
-    Advanced Customization
-    ~~~~~~~~~~~~~~~~~~~~~~
+    print("\n".join([str(x) for x in callback.results]))
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:448]: Oh this is interesting... is this an intended pattern where callbacks are stateful and accessed after training?

File: doc/source/train/user_guide.rst [hunk 8]
```diff
@@ -923,59 +874,59 @@ number of retries is configurable through the ``max_retries`` argument of the
 
 .. TODO.
 
-.. _train-profiling:
+.. .. _train-profiling:
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:877]: Add a note for this explaining why it's no longer supported (callbacks) and what the long term solution will be (checkpoints)

File: doc/source/train/user_guide.rst [hunk 9]
```diff
@@ -1033,7 +984,7 @@ Reproducibility
 .. tabbed:: PyTorch
 
     To limit sources of nondeterministic behavior, add
-    ``train.torch.enable_reproducibility()`` to the top of your training
+    ``train.torch.enable_reproducibility`` to the top of your training
     function. `
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:988]: Just noticed the extra backtick

```suggestion
    function.
```

File: doc/source/train/user_guide.rst [hunk 10]
```diff
@@ -1033,7 +984,7 @@ Reproducibility
 .. tabbed:: PyTorch
 
     To limit sources of nondeterministic behavior, add
-    ``train.torch.enable_reproducibility()`` to the top of your training
+    ``train.torch.enable_reproducibility`` to the top of your training
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:987]: Why remove the `()`?

Comment (Yard1) [reply doc/source/train/user_guide.rst:987]: Unintentional

File: doc/requirements-doc.txt [hunk 11]
```diff
@@ -29,6 +29,7 @@ tabulate
 uvicorn==0.16.0
 werkzeug
 wandb
+tensorflow
```
Comment (richardliaw) [inline doc/requirements-doc.txt:32]: pin?

Comment (Yard1) [reply doc/requirements-doc.txt:32]: Do we need to pin just for docs?

File: doc/source/train/user_guide.rst [hunk 12]
```diff
@@ -182,73 +183,87 @@ training.
...
+        )
 
 .. tabbed:: Horovod
 
     .. code-block:: python
 
-        from ray.train import Trainer
-        trainer = Trainer(backend="horovod", num_workers=2)
-
+        from ray.train.horovod import HorovodTrainer
```
Comment (amogkam) [inline doc/source/train/user_guide.rst:219]: Is it possible to run these code snippets in CI?

**Review by amogkam — COMMENTED** [review]
Thanks @Yard1!

My thoughts are that rather than just porting over the existing user guide, we should figure out what would be the best structure for AIR.

Some high level comments:
1. What is the lifecycle of a Trainer? What happens underneath the hood? This can be copied over from the current Trainer docstring
2.  We show basic Result usage, but do we want to show `metrics_dataframe` and `best_checkpoints` as well?
3. What are all the things I can do with RunConfig? For example cloud checkpointing is not possible with the old API but now can be enabled for training
4. How do Trainers work with Predictors? We should add a section linking to the relevant Predictor docs
5. Is there a developer guide on how I can implement my own Trainer/integrate with AIR? This can also be copied over from the existing docstring.
6. Are there good diagrams we can add to show how Trainer fits in with all the AIR components?

Also lot of these concepts are the same across Deep learning Trainers and non-DL Trainers. My preference is to have this user guide be global for all trainers and have a specific sub-section on Deep learning Trainers for the DL specific parts (showcasing the session API).

Let's discuss this offline some more, or if you have an idea on what you would like the structure to look like, would you want to open up a new PR with the docs skeleton and we can iterate on that?
cc @matthewdeng @richardliaw

**Comment by Yard1** [comment]
Thanks for the feedback @amogkam! I think we should work on this iteratively. Priority is removal of old APIs from the docs so we don't steer people the wrong way. I 100% agree we should have a better AIR guide, but that wasn't the intent of this PR. I would love to collaborate on that with you and @matthewdeng.

The doc code snippets can be ran in CI and I would like to do that in a followup (as mentioned in description).

File: doc/source/ray-air/package-ref.rst [hunk 13]
```diff
@@ -126,6 +126,19 @@ Configs
...
+.. _air-builtin-callbacks:
+
+Callbacks
+~~~~~~~~~
+
+Keras
+#####
+
+.. autoclass:: ray.air.callbacks.keras.Callback
+    :members:
```
Comment (amogkam) [inline doc/source/ray-air/package-ref.rst:139]: Do we want to also include the Tune callbacks here?

File: doc/source/train/api.rst [hunk 14]
```diff
@@ -4,8 +4,6 @@
 Ray Train API
 =============
 
-.. _train-api-trainer:
-
 Trainer
```
Comment (amogkam) [inline doc/source/train/api.rst:7]: Let's remove the deprecated Trainer APIs and replace with the AIR APIs?

File: doc/source/ray-air/package-ref.rst (hunk 13 above)
Comment (Yard1) [reply doc/source/ray-air/package-ref.rst:139]: Yes!

File: doc/source/train/api.rst (hunk 14 above)
Comment (amogkam) [reply doc/source/train/api.rst:7]: Can we just have the AIR APIs in both places (in AIR docs and Train docs)?

**Review by amogkam — COMMENTED** [review]
Thanks @Yard1. For the API references should we have them in both the AIR and the Train docs?

**Review by richardliaw — APPROVED** [review]

**Pull request merged**